pkg/planner/core/index_join_path.go

```diff
@@ -634,35 +633,29 @@ func getIndexJoinIntPKPathInfo(ds *logicalop.DataSource, innerJoinKeys, outerJoi
 
 // getBestIndexJoinInnerTaskByProp tries to build the best inner child task from ds for index join by the given property.
 func getBestIndexJoinInnerTaskByProp(ds *logicalop.DataSource, prop *property.PhysicalProperty,
-	opt *optimizetrace.PhysicalOptimizeOp, planCounter *base.PlanCounterTp) (base.Task, int64, error) {
+	planCounter *base.PlanCounterTp) (base.Task, int64, error) {
 	// the below code is quite similar from the original logic
 	// reason1: we need to leverage original indexPathInfo down related logic to build constant range for index plan.
 	// reason2: the ranges from TS and IS couldn't be directly used to derive the stats' estimation, it's not real.
 	// reason3: skyline pruning should not prune the possible index path which could feel the runtime EQ access conditions.
-	innerTSCopTask := buildDataSource2TableScanByIndexJoinProp(ds, prop)
-	if !innerTSCopTask.Invalid() {
-		planCounter.Dec(1)
-		if planCounter.Empty() {
-			// planCounter is counted to end, just return this one.
-			return innerTSCopTask, 1, nil
+	//
+	// here we build TableScan(TS) and IndexScan(IS) separately according to different index join prop is for we couldn't decide
+	// which one as the copTask here is better, some more possible upper attached operator cost should be
+	// considered, besides the row count, double reader cost for index lookup should also be considered as
+	// a whole, so we leave the cost compare for index join itself just like what it was before.
+	if prop.IndexJoinProp.TableRangeScan {
+		innerTSCopTask := buildDataSource2TableScanByIndexJoinProp(ds, prop)
+		if innerTSCopTask.Invalid() {
+			return base.InvalidTask, 0, nil
 		}
-	}
-	innerISCopTask := buildDataSource2IndexScanByIndexJoinProp(ds, prop)
-	if !innerISCopTask.Invalid() {
 		planCounter.Dec(1)
-		if planCounter.Empty() {
-			// planCounter is counted to end, just return this one.
-			return innerISCopTask, 1, nil
-		}
-	}
-	// if we can see the both, compare the cost.
-	leftIsBetter, err := compareTaskCost(innerTSCopTask, innerISCopTask, opt)
-	if err != nil {
-		return base.InvalidTask, 0, err
-	}
-	if leftIsBetter {
 		return innerTSCopTask, 1, nil
 	}
+	innerISCopTask := buildDataSource2IndexScanByIndexJoinProp(ds, prop)
+	if innerISCopTask.Invalid() {
+		return base.InvalidTask, 0, nil
+	}
+	planCounter.Dec(1)
```

Comment:
How about removing `planCounter` later? No one is using this, and it'll increase the complexity of our code.

Reply:
This comes from a nth plan-related test failure, and it was added. I will remove it systematically after that.